packages/core/certora/optimistic-asserter/specs/Asserter_Auxiliary.spec

```diff
@@ -0,0 +1,186 @@
+rule viewFuncsDontRevert(method f) 
+filtered{f -> f.isView} { // filters the methods to only view functions
+    env e;
+    require e.msg.value == 0;
+    calldataarg args;
+    f@withrevert(e, args);
+
+    assert !lastReverted;
+}
+
+/**************************************************
+ *                 "Who changed" Rules            *
+ **************************************************/
+// This type of rules calls any of the main contract's public/external methods
+// and checks whether a state variable was changed after the 
+// invocation of this function.
+
+rule whoChanged_cachedOracle(method f) 
+filtered{f -> !f.isView} { // We filter out view functions as they do not change state variables.
+    env e;
+    calldataarg args;
+    uint256 cachedOracle1 = cachedOracle();
+        f(e, args);
+    uint256 cachedOracle2 = cachedOracle();
+
+    assert cachedOracle1 == cachedOracle2;
+}
+
+
+rule whoChanged_burnedBondPercentage(method f)
+filtered{f -> !f.isView} {
+    env e;
+    calldataarg args;
+    uint256 burnPer1 = burnedBondPercentage();
+        f(e,args);
+    uint256 burnPer2 = burnedBondPercentage();
+
+    assert burnPer1 == burnPer2;
+}
+
+rule whoChanged_defaultCurrency(method f)
+filtered{f -> !f.isView} {
+    env e;
+    calldataarg args;
+    uint256 defaultCurrency1 = defaultCurrency();
+        f(e,args);
+    uint256 defaultCurrency2 = defaultCurrency();
+
+    assert defaultCurrency1 == defaultCurrency2;
+}
+
+rule whoChanged_settlementResolution(method f, bytes32 ID)
+filtered{f -> !f.isView} {
+    env e;
+    calldataarg args;
+    bool res1 = getAssertionSettlementResolution(ID);
+        f(e, args);
+    bool res2 = getAssertionSettlementResolution(ID);
+
+    assert res1 == res2;
+}
+
+rule whoChanged_assertionBond(method f, bytes32 ID)
+filtered{f -> !f.isView && !isMultiCall(f)} {
+    env e;
+    calldataarg args;
+    uint256 bond1 = getAssertionBond(ID);
+        f(e, args);
+    uint256 bond2 = getAssertionBond(ID);
+
+    assert bond1 == bond2;
+}
+
+rule whoChanged_assertionCurrency(method f, bytes32 ID)
+filtered{f -> !f.isView && !isMultiCall(f)} {
+    env e;
+    calldataarg args;
+    address currency1 = getAssertionCurrency(ID);
+        f(e, args);
+    address currency2 = getAssertionCurrency(ID);
+
+    assert currency1 == currency2;
+}
+
+/*************************************************
+*                Custom rules                    *
+**************************************************/
+// If calling 'assertTruth' twice, it cannot yield the same assertion ID.
+rule cannotAssertTruthTwiceForSameID() {
+    env e1;
+    env e2;
+    calldataarg args1; 
+    calldataarg args2;
+    bytes32 ID1 = assertTruth(e1, args1);
+    bytes32 ID2 = assertTruth(e2, args2);
+
+    // It's enough to check this rule for a single instance of the currency
+    require getAssertionCurrency(ID1) == testERC20;
+    require getAssertionCurrency(ID2) == testERC20;
```

Comment:
Why do we need to constrain this? Is this test stronger without this? Or does the test run forever if we don't constrain it?

Reply:
Generally, constraining is mainly useful if we expect the rule to fail, then we may consider only one possible option for the ERC20 identity.
The test is indeed stronger without this, but the constraint should help the prover run this faster.
I suppose that calling the function twice with two different currencies should yield different IDs. 
My purpose was to check all other parameters, but one can write another rule and check the currency dependency separately.

Reply:
Makes sense!